auth/services/oauth/holder.go

```diff
@@ -0,0 +1,139 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package oauth
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/auth/client/iam"
+	"github.com/nuts-foundation/nuts-node/auth/oauth"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/vcr/holder"
+	"github.com/nuts-foundation/nuts-node/vcr/pe"
+	"github.com/nuts-foundation/nuts-node/vcr/signature/proof"
+	"time"
+)
+
+var _ Holder = (*HolderService)(nil)
+
+var ErrNoCredentials = errors.New("no matching credentials")
+
+type HolderService struct {
+	strictMode        bool
+	httpClientTimeout time.Duration
+	httpClientTLS     *tls.Config
+	wallet            holder.Wallet
+}
+
+// NewHolder returns an implementation of Holder
+func NewHolder(wallet holder.Wallet, strictMode bool, httpClientTimeout time.Duration, httpClientTLS *tls.Config) *HolderService {
+	return &HolderService{
+		wallet:            wallet,
+		strictMode:        strictMode,
+		httpClientTimeout: httpClientTimeout,
+		httpClientTLS:     httpClientTLS,
+	}
+}
+
+func (v *HolderService) BuildPresentation(ctx context.Context, walletDID did.DID, presentationDefinition pe.PresentationDefinition, acceptedFormats map[string]map[string][]string, nonce string) (*vc.VerifiablePresentation, *pe.PresentationSubmission, error) {
+	// get VCs from own wallet
+	credentials, err := v.wallet.List(ctx, walletDID)
+	if err != nil {
+		return nil, nil, errors.New("failed to retrieve wallet credentials")
+	}
+
+	expires := time.Now().Add(time.Minute * 15) //todo
+	// build VP
+	submissionBuilder := presentationDefinition.PresentationSubmissionBuilder()
+	submissionBuilder.AddWallet(walletDID, credentials)
+	format := pe.ChooseVPFormat(acceptedFormats)
+	presentationSubmission, signInstructions, err := submissionBuilder.Build(format)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to build presentation submission: %w", err)
+	}
+	if signInstructions.Empty() {
+		return nil, nil, ErrNoCredentials
+	}
+
+	// todo: support multiple wallets
+	vp, err := v.wallet.BuildPresentation(ctx, signInstructions[0].VerifiableCredentials, holder.PresentationOptions{
+		Format: format,
+		ProofOptions: proof.ProofOptions{
+			Created:   time.Now(),
+			Challenge: &nonce,
+			Expires:   &expires,
+		},
+	}, &walletDID, false)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create verifiable presentation: %w", err)
+	}
+	return vp, &presentationSubmission, nil
```

Comment:
Thinking of cardinality here, or at least how the submission refers to the VPs. We should probably return a PEX envelope here? Since there could be multiple VPs (even if there's only 1 wallet), and how we should return the VP, depends the submission (flat referral to JSON object, indexed JSON array with 1 element, array with multiple elements).

Reply:
This would affect code on multiple places and can better be done in a separate PR.